### PR TITLE
Search in Xstate

### PR DESCRIPTION
**Pull request by spaceo**
This PR is showing how the search functionality with url param handling and infinite scroll can be implemnted in Xstate.

Things I like about this approach:
* The self [explaining flows and full control of states](https://github.com/danskernesdigitalebibliotek/dpl-go/blob/7f0323f261dc55ef16783471fd82d188c82f91c8/lib/machines/search/search.machine.ts) and data/context mutation
* [Cleaner components](https://github.com/danskernesdigitalebibliotek/dpl-go/blob/3cde3d1c813fc46197186a0b59caaad2e971d40a/components/pages/searchPageLayout/SearchPageLayout.tsx)
* Mulighed for at bruge middleware, som i det her tilfælde bliver brugt til [URL params](https://github.com/danskernesdigitalebibliotek/dpl-go/pull/34/files#diff-a3e8be467f7145bcab3236119f0b3d2d4bf9b4f0030d171a25a9d224f89f95c1R62)

Negative parts
* It is a matter of how you look at it. There is more lines of code in this approach but I like the full control of state transitions and data mutation
* There is a learning curve for newcomers to Xstate. But I think this work explains a lot of the principles.

### CODE REVIEW

File: components/shared/searchInput/SearchInput.tsx [hunk 1]
```diff
@@ -16,36 +18,34 @@ type SearchInputProps = {
 const SearchInput = ({ className, placeholder }: SearchInputProps) => {
   const inputRef = useRef<HTMLInputElement>(null)
   const router = useRouter()
-  const searchParams = useSearchParams()
-  const [queryString, setQueryString] = React.useState("")
+  const actor = useSearchMachineActor()
+  const currentQ = useSelector(actor, snapshot => {
```
Comment (ThomasGross) [inline components/shared/searchInput/SearchInput.tsx:22]: I dont like abbreviation / shortforms. Q could be hard to figure out what it is exactly. currentQuery is a much more descriptive name IMO.

In general i like descriptive code and variable names.

```suggestion
  const currentQuery = useSelector(actor, snapshot => {
```

File: lib/machines/search/queries.ts
Comment (ThomasGross) [inline lib/machines/search/queries.ts:1]: The functions inside this file are in a folder called "machines" but the functionality not really tied to a machine.

File: lib/machines/search/search.machine.setup.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      },
+    }),
+    emitFilterToggled: emit(({ event }) => ({
+      type: "filterToggled",
+      toggled: event,
+    })),
+    assignQToContext: assign({
+      currentQ: ({ event }) => event.q,
+    }),
+    resetFiltersIfNoQ: assign({
```
Comment (ThomasGross) [inline lib/machines/search/search.machine.setup.ts:43]:
```suggestion
    resetFiltersIfNoQuery: assign({
```

File: lib/machines/search/search.machine.setup.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+          ...selectedFilters,
+          [name]: [...(selectedFilters[filterName] ?? []), value],
+        }
+      },
+    }),
+    emitFilterToggled: emit(({ event }) => ({
+      type: "filterToggled",
+      toggled: event,
+    })),
+    assignQToContext: assign({
```
Comment (ThomasGross) [inline lib/machines/search/search.machine.setup.ts:40]:
```suggestion
    assignQueryToContext: assign({
```

File: lib/machines/search/search.machine.ts
Comment (ThomasGross) [inline lib/machines/search/search.machine.ts:1]: I dont understand the difference between the "search.machine.ts" and "search.machine.setup.ts" files

File: lib/machines/search/search.machine.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  id: "search",
+  initial: "bootstrap",
+  context: ({ input }) => ({
+    searchOffset: input.initialOffset ?? 0,
+    searchPageSize: input.searchPageSize,
+    facetLimit: input.facetLimit,
+    currentQ: input.q ?? "",
```
Comment (ThomasGross) [inline lib/machines/search/search.machine.ts:13]:
```suggestion
    currentQuery: input.q ?? "",
```

File: lib/machines/search/useSearchMachineActor.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  const url = new URL(window.location.href)
+  const {
+    toggled: { name: filterName, value: filterValue },
+  } = emittedEvent
+
+  if (url.searchParams.has(filterName, filterValue)) {
+    url.searchParams.delete(filterName, filterValue)
+  } else {
+    url.searchParams.append(filterName, filterValue)
+  }
```
Comment (ThomasGross) [inline lib/machines/search/useSearchMachineActor.tsx:32]: Will this work for multiple filters of the same type?

**Review by ThomasGross — CHANGES_REQUESTED** [review]

File: components/shared/searchInput/SearchInput.tsx (hunk 1 above)
Comment (spaceo) [reply components/shared/searchInput/SearchInput.tsx:22]: 👍

File: components/pages/searchPageLayout/helper.ts [hunk 6]
```diff
@@ -57,3 +60,23 @@ export const getNextPageParamsFunc = (
...
+  const actor = useSearchMachineActor()
+  const q = useSelector(actor, snapshot => {
+    return snapshot.context.currentQ
+  })
+  const data = useSelector(actor, snapshot => {
+    const { facetData: facets, searchData: search } = snapshot.context
+    return { facets, search }
+  })
+  const isLoadingFacets =
+    !data.facets || actor.getSnapshot().matches({ filteringAndSearching: "filter" })
```
[the remark this answers is not in the document]
Comment (spaceo) [reply components/pages/searchPageLayout/helper.ts:74]: Investigate this

File: lib/machines/search/search.machine.setup.ts (hunk 2 above)
Comment (spaceo) [reply lib/machines/search/search.machine.setup.ts:43]: 👍

File: lib/machines/search/search.machine.setup.ts (hunk 3 above)
Comment (spaceo) [reply lib/machines/search/search.machine.setup.ts:40]: 👍

**Pull request merged**